sasdata/transforms/NDrebin.py

```diff
@@ -0,0 +1,522 @@
+    binned_data_errs : 
+        has size num_bins and is NDimensional, contains
+        the propagated errors of the binned_data
+    bins_list : 
+        is a list of 1D vectors, is similar to bin_centers_list,
+        but instead contains the edges of the bins, so it is 1 longer
+        in each dimension
+    step_size : 
+        is a list of Ndims numbers, contains the step size
+        along each dimensino
+    num_bins : 
+        is a list of Ndims numbers, contains the number
+        of bins along each dimension
+
+    Methods
+    -------
+    run(self):
+        Bin the data into the defined bins.
+
+    Typical usage
+    -------------
+    .. code-block::
+    # test syntax 1
+    Ndims = 4
+    Nvals = int(1e4)
+    qmat = np.random.rand(Ndims, Nvals)
+    Imat = np.random.rand(Nvals)
+
+    rebin = NDRebin(Imat, qmat,
+        step_size=0.1*np.random.rand(Ndims)+0.05,
+        lower=0.1*np.random.rand(Ndims)+0.0,
+        upper=0.1*np.random.rand(Ndims)+0.9)
+    rebin.run()
+
+    Ibin = rebin.binned_data
+    qbin = rebin.bin_centers_list
+
+
+    # test syntax 2
+    Ndims = 2
+    Nvals = int(1e4)
+    qmat = np.random.rand(Ndims, 100, Nvals)
+    Imat = np.random.rand(100, Nvals)
+    Imat_errs = np.random.rand(100, Nvals)
+
+    rebin = NDRebin(Imat, qmat,
+        data_errs = Imat_errs,
+        num_bins=[10,20],
+        axes = np.eye(2),
+        fractional=True)
+    rebin.run()
+
+    Ibin = rebin.binned_data
+    qbin = rebin.bin_centers_list
+    Ibin_errs = rebin.binned_data_errs
+    bins_list = rebin.bins_list
+    step_size = rebin.step_size
+    num_bins = rebin.num_bins
+    """
+
+    def __init__(
+        self,
+        data: Quantity[ArrayLike],
+        coords: Quantity[ArrayLike],
+        data_errs: Quantity[ArrayLike] | None = None,
+        axes: ArrayLike | None = None,
+        upper: ArrayLike | None = None,
+        lower: ArrayLike | None = None,
+        step_size: ArrayLike | None = None,
+        num_bins: ArrayLike | None = None,
+        fractional: bool = False,
+        normalize: bool = True,
+    ):
+        self.data = data
+        self.coords = coords
+        self.data_errs = data_errs
+        self.axes = axes
+        self.upper = upper
+        self.lower = lower
+        self.step_size = step_size
+        self.num_bins = num_bins
+        self.fractional = fractional
+        self.normalize = normalize
+
+        # Internal attributes initialised later
+        self.Nvals: int | None = None
+        self.Ndims: int | None = None
+        self.data_flat = None
+        self.errors_flat = None
+        self.coords_flat = None
+        self.bins_list = None
+        self.bin_centers_list = None
+        self.bin_inds = None
+        self.binned_data = None
+        self.binned_data_errs = None
+        self.n_samples = None
+
+        self._prepared = False   # flag to avoid double-prepare
+
+    def __call__(self):
+        self.run()
+        return self.binned_data
```

Comment:
Why does the this routine return `self.binned_data`? Don't you want the functionality to be the same as `self.run`?